core/src/apps/bitshares/base/types.py

```diff
@@ -0,0 +1,375 @@
+
+    @property
+    def address(self) -> bytes:
+        sha_digest = sha512(self.pub_key).digest()
+        ripemd_digest = ripemd160(sha_digest).digest()
+        return ripemd_digest
+
+    # TODO
+    def _is_compressed(self, pub_key) -> bool:
+        return pub_key.startswith("BTS")
+
+    # TODO
+    def _is_raw(self, pub_key) -> bool:
+        return len(pub_key) == 33
+
+    def __lt__(self, other) -> bool:
+        return self.address < other.address
+
+
+class ObjectId(VarInt32):
+    def __init__(self, msg) -> None:
+        self.data = self._readable = msg
+
+    def parse(self) -> self:
+        if self.data is None:
+            raise ValueError
+
+        self.data = self._get_instance(self.data)
+        return self
+
+    def __str__(self) -> str:
+        # TODO currently returns full ID, probably return obj_id only
+        return self._readable
+
+    def _get_instance(self, oid) -> int:
+        try:
+            return int(oid.split(".")[2])
+        except:
+            raise wire.DataError("ObjectId {} wrong.".format(oid))
+
+
+class VoteId(Uint32):
+    def __init__(self, msg) -> None:
+        self.data = self.readable = msg
+
+    def parse(self) -> None:
+        super().parse()
+
+        self.data = self._serialize(self.data)
+        return self
+
+    def _serialize(self, vid: str) -> int:
+        try:
+            _type, _instance = vid.split(":")
+            return int(_type) & 0xFF | int(_instance) << 8
+        except:
+            raise wire.DataError("Bad VoteId format.")
+
+    def __str__(self) -> str:
+        return self._readable
+
+
+class OperationId(Uint8):
+    def __init__(self, id: int) -> None:
+        self.data = id
+
+
+class EmptyExtension(Uint8):
+    def __init__(self, msg) -> None:
+        msg = 0
+        super().__init__(msg)
+
+
+class Memo(Node):
+    def __init__(self, msg: BitSharesMemo) -> None:
+        self.data = msg
+
+        self.nodes = OrderedDict(
+            [
+                ("sender", PublicKey),
+                ("receiver", PublicKey),
+                ("nonce", Uint64),
+                ("message", Bytes),
+            ]
+        )
+
+
+class Asset(Node):
+    def __init__(self, msg: BitSharesAsset) -> None:
+        self.data = msg
+        self.nodes = OrderedDict([("amount", Uint64), ("asset_id", ObjectId),])
+
+
+class AccountAuthority(Node):
+    def __init__(self, msg: BitSharesAccountAuthority) -> None:
+        self.data = msg
+        self.nodes = OrderedDict([("account", ObjectId), ("weight", Uint16),])
+
+
+class KeyAuthority(Node):
+    def __init__(self, msg: BitSharesKeyAuthority) -> None:
+        self.data = msg
+        self.nodes = OrderedDict([("key", PublicKey), ("weight", Uint16),])
+
+
+class AddressAuthority(Node):
+    def __init__(self, msg: BitSharesAddressAuthority) -> None:
+        super().__init__(msg)
+        self.nodes = OrderedDict(
+            [("address", Bytes), ("weight", Uint16),]  # TODO rly bytes?
```

Comment:
YEs, address is in bytes .. not used anywhere in bitshares, could aswell remove it.